Split constants file into modules, and split exported constants into modules

daboross opened the pull request:
This addresses the last item in and closes #88.

The first commit here splits `constants.rs` into multiple files based on type of constant, and the last commit adds a bunch of modules which organize the exported constants based on the area of the game they affect. Probably way too many distinct modules, we can work on making that fewer.

The middle two commits are fairly small fixes semi-related to the above.

ASalvail commented on src/constants.rs:
```diff
@@ -59,1279 +59,157 @@
+            LAB_MINERAL_CAPACITY, LAB_REACTION_AMOUNT, LAB_UNBOOST_ENERGY, LAB_UNBOOST_MINERAL,
+            MINERAL_DENSITY_CHANGE, MINERAL_RANDOM_FACTOR, MINERAL_REGEN_TIME,
+        },
+        small_enums::Density,
```
Maybe include `ResourceType` here as it mostly pertains to minerals.

ASalvail commented on src/constants.rs:
```diff
@@ -59,1279 +59,157 @@
+pub mod source {
+    pub use super::numbers::{
+        ENERGY_REGEN_TIME, INVADERS_ENERGY_GOAL, SOURCE_ENERGY_CAPACITY,
+        SOURCE_ENERGY_KEEPER_CAPACITY, SOURCE_ENERGY_NEUTRAL_CAPACITY,
```
Might want to add `ENERGY_DECAY` here.

daboross replied:
I _believe_ that constant is more related to how dropped resources behave than `Source`s. Unless you'd propose merging the two modules? I can get behind that.

ASalvail replied:
Well, beside fights, I expect to care about dropped energy when dealing with sources... Just a thought.

daboross replied:
Ah, that's a good point. I'll stick it in here too!

ASalvail commented on src/constants.rs:
```diff
@@ -59,1279 +59,157 @@
-    }
+/// Re-export of all constants related to `Resource`s.
+pub mod resource {
+    pub use super::{numbers::ENERGY_DECAY, types::ResourceType};
```
Might be worth adding `InterShardResource` here.

daboross replied:
I was thinking of `resource` mainly being about dropped resources / the `Resource` type. I think `IntershardResourceType`, while named similarly, is unrelated?

ASalvail replied:
Only problem is that it doesn't appear anywhere. Although, I guess it's still in `constants`...

daboross replied:
true, it doesn't really have another.. place. I mostly figured it'd be useful alone, like `Terrain` or `Direction`. Maybe all of those should have some place too, though? I'm not sure.

daboross commented:
Rebased on master, added `ResourceType` to `minerals` module, and added links to submodule documentation:
![screenshot-20190822-192511](https://user-images.githubusercontent.com/1152146/63562467-94521000-c512-11e9-8d31-5429ae34c751.png)

I like the `resources` module even though it's... a bit bare. But I can see it being unfair to have it not others like `link`...

With `portal`, I separated it out mostly since I was thinking of its nature as a non-player-created and non-destroyable structure. Most of the things in `structure` seem useful for creating and destroying structures, and portals usually require fairly unique consideration. Though I guess everything matters for pathfinding, and that might be an arbitrary split...

ASalvail commented:
@Dessix @PatrickVdWillik any thoughts? It's meant as a convenience.

ASalvail commented:
Btw, it looks very good.  We're starting to have documentation that doesn't suck 🎉
